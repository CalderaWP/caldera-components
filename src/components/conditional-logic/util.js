/**
 * Get the current values for prepareConfigFields
 *
 * @param configFields
 */
export const reduceConfigFieldsToValues = (configFields) => {
	let values = {};
	configFields.forEach(field => {
		values[field.ID] = field.value
			? field.value
			: field.default
				? field.default
				: null;
	});
	return values;
};
