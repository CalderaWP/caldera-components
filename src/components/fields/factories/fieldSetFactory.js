import {fieldFactory} from './fieldFactory';

/**
 * Given an array of field configs, generates and array of field controls
 *
 * Should be used for a group of settings, such as all settings for a processor.
 *
 * @param {Array} fields Field configs
 * @returns {Array}
 */
export const fieldSetFactory = (fields) =>{
	let out = [];
	Object.values(fields).forEach( field => {
		out.push(fieldFactory(field));
	});
	return out;
};