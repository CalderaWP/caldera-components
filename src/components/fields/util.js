/**
 * Create the aria-describedby attribute or id attribute for the describing element
 *
 * @param {String} id Id attribute of input being described
 * @param {String} help Optional. Help text. If empty return is null.
 * @returns {String|null}
 */
export const  ariaDescribedbyAttr = (id, help = '') => {
	let ariaDescribedby = null;
	if (help.length) {
		ariaDescribedby = `${id}-description`;
	}
	return ariaDescribedby;
};


/**
 * Get HTML5 input types that are valid
 * @returns {string[]}
 */
export function getHtmlInputTypes() {
	return [
		'text',
		'email',
		'number',
		'date',
		'datetime',
		'password',
		'submit',
		'reset',
		'checkbox',
		'hidden'
	];
}

/**
 * Checks if a given input type is an acceptable HTML5 input type
 *
 * @param {String} type
 * @returns {boolean}
 */
export function isValidHtml5type(type) {
	return getHtmlInputTypes().includes(type);
}


/**
 * Remove a value from an array if present, if not present, add it
 *
 * @param {String|number} value Value to add or remove
 * @param {Array} array Array to mutate
 * @return {*}
 */
export const addOrRemoveFromArray = (value,array) => {
	const index = array.indexOf(value);
	if (index !== -1){
		array.splice(index, 1);
	}else{
		array.push( value );
	}
	return array;
};

/**
 * Cast a boolean or boolean like to a true or false
 *
 * @param Mixed} value Value to cast
 * @return {boolean}
 */
export const toBoolean =(value) => {
	switch(value){
	case true:
	case 'true':
	case 1:
	case '1':
	case 'on':
	case 'yes':
		return true;
	default:
		return false;
	}
};