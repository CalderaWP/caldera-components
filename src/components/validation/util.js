import getValidatorsFromConfigField from './getValidatorsFromConfigField';
import isValid from './isValid';
import isValidOrEmpty from './isValidOrEmpty';

/**
 * Find the value of a field by ID
 *
 * Returns null if field is not in collection
 *
 * @param {String} fieldId Field to look for
 * @param {Object} fieldValues Values to look for field in
 * @return {String|Array|boolean|number|null}
 */
export const findFieldValueInFieldValues = (fieldId, fieldValues) => {
	if (fieldValues.hasOwnProperty(fieldId)) {
		return fieldValues[fieldId];
	}

	return null;
};

/**
 * Helper function to curry validator functions
 *
 * Validators expect 1 value, not an object of values, this function solves that.
 *
 * @param {String} fieldId Field to look for
 * @param {Function} validator Validator function
 * @return {function(*=): *}
 */
function curryValidator(fieldId, validator) {
	return (fieldValues) => {
		return validator(findFieldValueInFieldValues(fieldId, fieldValues));
	};
}

/**
 * Create validation function based on type and required
 *
 * @param {Object} configField Field to create validations for
 * @param {Array} validators Current array of validators
 * @param {string} validationType Type of validators to add.
 * @return {Array}
 */
function addValidatorsForType(configField, validators, validationType) {
	if (configField.isRequired) {
		validators.push(
			curryValidator(
				configField.ID, isValid[validationType]
			)
		);
	} else {
		validators.push(
			curryValidator(
				configField.ID, isValidOrEmpty[validationType]
			)
		);
	}

	return validators;
}

/**
 * Adds sensible, automatic validation to fields -- if no validators have been added yet.
 *
 * @param {Object} configField
 * @return {*}
 */
export const addAutomaticValidators = (configField) => {
	let validators = getValidatorsFromConfigField(configField);
	if (0 !== validators.length) {
		return configField;
	}

	switch (configField.type) {
	case 'input':
		if( [
			'email',
			'url',
			'date',
			'number',
		].includes( configField.inputType ) ){
			validators = addValidatorsForType(configField, validators, configField.inputType);
		}else if( 'text' === configField.inputType){
			validators = addValidatorsForType(configField, validators, 'string');

		}

		else{
			validators = addValidatorsForType( configField, validators, 'anything' );
		}

		break;
	case 'select':
		validators = addValidatorsForType( configField, validators, 'stringOrNumber' );
		break;
	}


	return {
		...configField,
		validators
	};


};
