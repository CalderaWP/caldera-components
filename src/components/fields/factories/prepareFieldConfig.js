import {fieldGroupPropTypes} from '../propTypes';
import {isValidHtml5type} from '../util';

/**
 * Validates field configurations
 *
 * NOTE: This is called internally by factory. You probably do not want to call this.
 *
 * @param {Object}fieldArgs
 * @returns {*}
 */
export const prepareFieldConfig = (fieldArgs) => {
	function pick(obj, keys) {
		return keys.map(k => k in obj ? {[k]: obj[k]} : {})
			.reduce((res, o) => Object.assign(res, o), {});
	}

	switch( fieldArgs.type ){
	case 'text':
	case 'number':
	default:
		fieldArgs.inputType = isValidHtml5type(fieldArgs.type) ? fieldArgs.type : 'text';
		fieldArgs.type = 'input';
		break;
	}
	if( fieldArgs.hasOwnProperty('desc') ){
		fieldArgs.help = 'desc';
	}
	return pick(fieldArgs, Object.keys(fieldGroupPropTypes));
};