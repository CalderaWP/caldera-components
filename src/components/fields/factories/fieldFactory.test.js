import {
	fieldFactory,
} from './fieldFactory';
import {prepareFieldConfig} from './prepareFieldConfig';
import {fieldSetFactory} from './fieldSetFactory';
import {getHtmlInputTypes} from '../util';
import { mount } from 'enzyme';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import renderer from 'react-test-renderer';
Enzyme.configure({ adapter: new Adapter() });

const textFieldConfig = {
	'id': 'cf-convertkit-tags',
	'label': 'Tags',
	'desc': 'Comma separated list of tags.',
	'type': 'text',
	'description': false
};

const hiddenFieldConfig = {
	'id': 'cf-convertkit-sequence-id',
	'type': 'hidden',
	'label': 'Sequence ID',
	'description': false
};

const configFields = [
	{
		'id': 'cf-convertkit-apikey', 'label': 'API Key', 'type': 'text'
	},
	{
		'id': 'pid-1',
		'label': 'Sequence',
		'type': 'dropdown',
		'options': ['-- Select A ConvertKit Sequence --'],
		'desc': 'ConvertKit sequence to add subscriber to. Sequences are also referred to as courses.',
		'description': false,
		'extra_classes': 'field-Something is wrong',
		'magic': false
	},
	{
		'id': 'cf-convertkit-sequence-id',
		'type': 'hidden',
		'label': 'Sequence ID',
		'description': false
	},
	hiddenFieldConfig,
	{
		'id': 'cf-convertkit-email',
		'label': 'Email Address',
		'desc': 'Subscriber email address.',
		'type': 'advanced',
		'allow_types': ['email'],
		'description': true,
		'magic': false
	},
	{
		'id': 'cf-convertkit-name',
		'label': 'Name',
		'type': 'text',
		'desc': 'Subscriber name.',
		'description': true
	},
	textFieldConfig
];

const genericHandler = () => {};
configFields.map(config => {
	config.onValueChange = genericHandler;
});

describe( 'Factories', () => {
	describe('Field factory', () => {

		it( 'validates type arg, setting inputType arg', () => {
			let config = {
				...textFieldConfig,
				type: 'email',
				onValueChange: genericHandler
			};
			expect( config.type ).toBe('email');
			config = prepareFieldConfig(config);
			expect( config.inputType ).toBe('email');
		});


		describe( 'Sets inputType arg in config', () => {
			getHtmlInputTypes().forEach((type) => {
				it(`type arg with value of ${type} sets inputType arg`, () => {
					const config = prepareFieldConfig({
						...textFieldConfig,
						type,
						onValueChange: genericHandler
					});
					expect( config.inputType ).toBe( type );
				});
			});

		});

		it( 'Creates inputs', () => {
			const component = fieldFactory({
				...textFieldConfig,
				onValueChange:genericHandler,
			});
			expect( component.type ).toBe( 'div' );
		});

		describe('Works for all HTML5 input types via inputType prop', () => {
			getHtmlInputTypes().forEach((type) => {
				it(`inputType prop of ${type} works`, () => {
					let config = {
						...textFieldConfig,
						type,
						onValueChange: genericHandler
					};
					const wrapper = mount(
						fieldFactory(config)
					);
					expect(wrapper.find('input').prop('type')).toBe(type);
				});

			});
		});

	});

	describe( 'Field set factory', () => {
		it( 'Creates the right number of elements', () => {
			const components = fieldSetFactory(configFields);
			expect( components ).toHaveLength(configFields.length);
		});

		it( 'Renders with elements', () => {
			const components = fieldSetFactory(configFields);
			const component = renderer.create(
				<div>
					{Array.from(components).map((field,i) => {
						return React.createElement(
							'div', {
								key: i,
								className: `f-${i}`
							},
							field
						);
					})}
				</div>
			);
			expect(component.toJSON()).toMatchSnapshot();
		});

		it( 'Creates the elements', () => {
			const components = fieldSetFactory(configFields);
			const wrapper = mount(
				<div>
					{Array.from(components).map((field,i) => {
						return React.createElement(
							'div', {
								key: i,
								className: `f-${i}`
							},
							field
						);
					})}
				</div>
			);
			expect( wrapper.children()).toHaveLength(configFields.length);
			expect( wrapper.find('.f-1')).toHaveLength(1);
			expect( wrapper.find('.f-1').text()).toBe('Sequence');
		});


	});
});
