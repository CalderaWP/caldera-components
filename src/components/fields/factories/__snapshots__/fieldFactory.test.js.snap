// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Factories Field set factory Renders with elements 1`] = `
<div>
  <div
    className="f-0"
  >
    <div
      className="caldera-config-field"
    >
      <label
        htmlFor="cf-convertkit-apikey"
      >
        API Key
      </label>
      <input
        aria-describedby={null}
        className="field-config field-config"
        disabled={false}
        id="cf-convertkit-apikey"
        onChange={[Function]}
        required={false}
        type="text"
        value={undefined}
      />
    </div>
  </div>
  <div
    className="f-1"
  >
    <div
      className="caldera-config-field"
    >
      <label
        htmlFor="pid-1"
      >
        Sequence
      </label>
      <select
        className="field-config field-config"
        disabled={false}
        id="pid-1"
        onChange={[Function]}
        value={undefined}
      >
        <option
          className="caldera-config-option"
          value={undefined}
        />
      </select>
    </div>
  </div>
  <div
    className="f-2"
  >
    <div
      className="caldera-config-field"
    >
      <label
        htmlFor="cf-convertkit-sequence-id"
      >
        Sequence ID
      </label>
      <input
        aria-describedby={null}
        className="field-config field-config"
        disabled={false}
        id="cf-convertkit-sequence-id"
        onChange={[Function]}
        required={false}
        type="text"
        value={undefined}
      />
    </div>
  </div>
  <div
    className="f-3"
  >
    <div
      className="caldera-config-field"
    >
      <label
        htmlFor="cf-convertkit-sequence-id"
      >
        Sequence ID
      </label>
      <input
        aria-describedby={null}
        className="field-config field-config"
        disabled={false}
        id="cf-convertkit-sequence-id"
        onChange={[Function]}
        required={false}
        type="text"
        value={undefined}
      />
    </div>
  </div>
  <div
    className="f-4"
  >
    <div
      className="caldera-config-field"
    >
      <label
        htmlFor="cf-convertkit-email"
      >
        Email Address
      </label>
      <input
        aria-describedby={null}
        className="field-config field-config"
        disabled={false}
        id="cf-convertkit-email"
        onChange={[Function]}
        required={false}
        type="text"
        value={undefined}
      />
      <p
        className="description"
        id="null"
      />
    </div>
  </div>
  <div
    className="f-5"
  >
    <div
      className="caldera-config-field"
    >
      <label
        htmlFor="cf-convertkit-name"
      >
        Name
      </label>
      <input
        aria-describedby={null}
        className="field-config field-config"
        disabled={false}
        id="cf-convertkit-name"
        onChange={[Function]}
        required={false}
        type="text"
        value={undefined}
      />
      <p
        className="description"
        id="null"
      />
    </div>
  </div>
  <div
    className="f-6"
  >
    <div
      className="caldera-config-field"
    >
      <label
        htmlFor="cf-convertkit-tags"
      >
        Tags
      </label>
      <input
        aria-describedby={null}
        className="field-config field-config"
        disabled={false}
        id="cf-convertkit-tags"
        onChange={[Function]}
        required={false}
        type="text"
        value={undefined}
      />
    </div>
  </div>
</div>
`;
